File: go/wrap.go
package mdtable

import (
	"io/fs"
	"os"
	"path/filepath"
	"strings"
)

// WrapString reflows prose paragraphs in s to width display columns and returns
// the result. It is pure and side-effect free: fenced code blocks, tables,
// headings, lists, blockquotes, thematic breaks, HTML blocks, indented code, and
// YAML front matter pass through untouched, and the original newline style (LF
// vs CRLF) and trailing-newline state are preserved. A width below 1 disables
// wrapping (each paragraph collapses to a single line).
func WrapString(s string, width int) string {
	lines := splitLines(s)

	out := make([]logicalLine, 0, len(lines))
	var para []logicalLine // accumulated prose lines awaiting a flush

	flush := func() {
		if len(para) == 0 {
			return
		}
		out = append(out, wrapParagraph(para, width)...)
		para = para[:0]
	}

	inFence := false
	var fenceMarker byte
	var fenceLen int
	inFrontMatter := false

	for i := 0; i < len(lines); i++ {
		line := lines[i]
		trimmed := strings.TrimSpace(line.text)

		// YAML front matter: a leading `---` on the very first line opens a block
		// that passes through verbatim until its closing `---`/`...`.
		if i == 0 && trimmed == "---" {
			inFrontMatter = true
			out = append(out, line)
			continue
		}
		if inFrontMatter {
			out = append(out, line)
			if trimmed == "---" || trimmed == "..." {
				inFrontMatter = false
			}
			continue
		}

		// Code fences pass through and toggle fence state (reusing the aligner's
		// CommonMark-aware detection).
		if !indentedTooFar(line.text) {
			if !inFence {
				if marker, n, ok := openingFenceToken(trimmed); ok {
					flush()
					inFence, fenceMarker, fenceLen = true, marker, n
					out = append(out, line)
					continue
				}
			} else if isClosingFence(trimmed, fenceMarker, fenceLen) {
				inFence = false
				out = append(out, line)
				continue
			}
		}
		if inFence {
			out = append(out, line)
			continue
		}

		// A pending paragraph immediately followed by a setext underline is a
		// heading; emit both verbatim so the underline keeps matching. (Checked
		// before prose classification because "===" is not otherwise a block.)
		if len(para) > 0 && isSetextUnderline(line.text) {
			out = append(out, para...)
			para = para[:0]
			out = append(out, line)
			continue
		}

		if isProse(line.text) {
			para = append(para, line)
			continue
		}

		// Any non-prose line ends the current paragraph and passes through.
		flush()
		out = append(out, line)
	}
	flush()

	var b strings.Builder
	for _, line := range out {
		b.WriteString(line.text)
		b.WriteString(line.eol)
	}
	return b.String()
}

// WrapDirectory walks root recursively and wraps every *.md file in place to
// width columns, rewriting only files whose content actually changes and
// preserving file mode. It returns the paths of the files it rewrote, in walk
// order.
func WrapDirectory(root string, width int) ([]string, error) {
	var changed []string
	err := filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() || !strings.EqualFold(filepath.Ext(path), ".md") {
			return nil
		}
		data, err := os.ReadFile(path)
		if err != nil {
			return err
		}
		wrapped := WrapString(string(data), width)
		if wrapped == string(data) {
			return nil
		}
		info, err := d.Info()
		if err != nil {
			return err
		}
		if err := os.WriteFile(path, []byte(wrapped), info.Mode().Perm()); err != nil {
			return err
		}
		changed = append(changed, path)
		return nil
	})
	return changed, err
}

// wrapParagraph greedily packs the words of a prose paragraph into lines no wider
// than width display columns, never splitting a single word. Newly created lines
// reuse the paragraph's first source eol; the final line keeps the paragraph's
// last source eol so trailing-newline and CRLF state survive.
func wrapParagraph(para []logicalLine, width int) []logicalLine {
	var words []string
	for _, line := range para {
		words = append(words, strings.Fields(line.text)...)
	}
	if len(words) == 0 {
		return para
	}

	firstEol := para[0].eol
	lastEol := para[len(para)-1].eol

	var texts []string
	if width < 1 {
		texts = []string{strings.Join(words, " ")}
	} else {
		cur := words[0]
		curW := dispWidth(cur)
		for _, w := range words[1:] {
			ww := dispWidth(w)
			if curW+1+ww <= width {
				cur += " " + w
				curW += 1 + ww
			} else {
				texts = append(texts, cur)
				cur, curW = w, ww
			}
		}
		texts = append(texts, cur)
	}

	wrapped := make([]logicalLine, len(texts))
	for i, text := range texts {
		eol := firstEol
		if i == len(texts)-1 {
			eol = lastEol
		}
		wrapped[i] = logicalLine{text: text, eol: eol}
	}
	return wrapped
}

// isProse reports whether a line is ordinary paragraph text — i.e. it has
// content and is none of the block constructs that must pass through untouched.
// Fenced code, indented code, and front matter are handled by the caller's state
// machine before this is consulted.
func isProse(line string) bool {
	if strings.TrimSpace(line) == "" {
		return false
	}
	if indentedTooFar(line) {
		return false
	}
	if isAtxHeading(line) || isBlockquote(line) || isListItem(line) ||
		isThematicBreak(line) || isHTMLBlockStart(line) || isPipeRow(line) {
		return false
	}
	return true
}

// leadingSpaces returns the count of leading ASCII spaces, capped at the 4 that
// matter for CommonMark block recognition.
func leadingSpaces(line string) int {
	n := 0
	for n < len(line) && n < 4 && line[n] == ' ' {
		n++
	}
	return n
}

// isAtxHeading reports whether a line is an ATX heading (1-6 '#' then a space or
// end of line), indented at most three columns.
func isAtxHeading(line string) bool {
	if indentedTooFar(line) {
		return false
	}
	s := line[leadingSpaces(line):]
	hashes := 0
	for hashes < len(s) && s[hashes] == '#' {
		hashes++
	}
	if hashes < 1 || hashes > 6 {
		return false
	}
	return hashes == len(s) || s[hashes] == ' ' || s[hashes] == '\t'
}

// isBlockquote reports whether a line begins a block quote (a '>' at most three
// columns in).
func isBlockquote(line string) bool {
	if indentedTooFar(line) {
		return false
	}
	s := line[leadingSpaces(line):]
	return len(s) > 0 && s[0] == '>'
}

// isListItem reports whether a line begins a bullet or ordered list item.
func isListItem(line string) bool {
	if indentedTooFar(line) {
		return false
	}
	s := line[leadingSpaces(line):]
	if s == "" {
		return false
	}
	if c := s[0]; c == '-' || c == '*' || c == '+' {
		return len(s) == 1 || s[1] == ' ' || s[1] == '\t'
	}
	digits := 0
	for digits < len(s) && digits < 9 && s[digits] >= '0' && s[digits] <= '9' {
		digits++
	}
	if digits == 0 || digits >= len(s) {
		return false
	}
	if s[digits] != '.' && s[digits] != ')' {
		return false
	}
	rest := digits + 1
	return rest == len(s) || s[rest] == ' ' || s[rest] == '\t'
}

// isThematicBreak reports whether a line is a thematic break: three or more of a
// single '-', '*', or '_' with only spaces between them.
func isThematicBreak(line string) bool {
	if indentedTooFar(line) {
		return false
	}
	s := strings.TrimSpace(line)
	if len(s) < 3 {
		return false
	}
	marker := s[0]
	if marker != '-' && marker != '*' && marker != '_' {
		return false
	}
	count := 0
	for i := 0; i < len(s); i++ {
		switch s[i] {
		case marker:
			count++
		case ' ':
		default:
			return false
		}
	}
	return count >= 3
}

// isSetextUnderline reports whether a line is a setext heading underline: a run
// of only '=' or only '-', indented at most three columns.
func isSetextUnderline(line string) bool {
	if indentedTooFar(line) {
		return false
	}
	s := strings.TrimRight(line[leadingSpaces(line):], " \t")
	if s == "" {
		return false
	}
	marker := s[0]
	if marker != '=' && marker != '-' {
		return false
	}
	for i := 0; i < len(s); i++ {
		if s[i] != marker {
			return false
		}
	}
	return true
}

// isHTMLBlockStart reports whether a line begins an HTML block (a '<' at most
// three columns in). Such lines are left untouched.
func isHTMLBlockStart(line string) bool {
	if indentedTooFar(line) {
		return false
	}
	s := line[leadingSpaces(line):]
	return len(s) > 0 && s[0] == '<'
}

File: rust/src/wrap.rs
//! Reflows markdown prose paragraphs to a display width. The Rust counterpart of
//! the Go `mdtable` wrapper; both are verified against the shared `testdata/wrap`
//! fixtures. Reuses the aligner's line, fence, and table primitives from the
//! crate root.

use std::borrow::Cow;
use std::fs;
use std::io;
use std::path::{Path, PathBuf};

use crate::common::{
    disp_width, indented_too_far, is_closing_fence, is_markdown, is_pipe_row, opening_fence_token,
    split_lines, LogicalLine,
};

/// Reflows prose paragraphs in `input` to `width` display columns and returns the
/// result. Pure and side-effect free: fenced code blocks, tables, headings,
/// lists, blockquotes, thematic breaks, HTML blocks, indented code, and YAML
/// front matter pass through untouched, and the original newline style (LF vs
/// CRLF) and trailing-newline state are preserved. A `width` of 0 disables
/// wrapping (each paragraph collapses to a single line).
pub fn wrap_str(input: &str, width: usize) -> String {
    let lines = split_lines(input);

    let mut out: Vec<(Cow<str>, &str)> = Vec::with_capacity(lines.len());
    let mut para: Vec<LogicalLine> = Vec::new();

    let mut in_fence = false;
    let mut fence_marker = 0u8;
    let mut fence_len = 0usize;
    let mut in_front_matter = false;

    for (i, line) in lines.iter().enumerate() {
        let trimmed = line.text.trim();

        // YAML front matter: a leading `---` on the very first line opens a block
        // that passes through verbatim until its closing `---`/`...`.
        if i == 0 && trimmed == "---" {
            in_front_matter = true;
            out.push((Cow::Borrowed(line.text), line.eol));
            continue;
        }
        if in_front_matter {
            out.push((Cow::Borrowed(line.text), line.eol));
            if trimmed == "---" || trimmed == "..." {
                in_front_matter = false;
            }
            continue;
        }

        // Code fences pass through and toggle fence state (reusing the aligner's
        // CommonMark-aware detection).
        if !indented_too_far(line.text) {
            if !in_fence {
                if let Some((marker, n)) = opening_fence_token(trimmed) {
                    flush(&mut out, &mut para, width);
                    in_fence = true;
                    fence_marker = marker;
                    fence_len = n;
                    out.push((Cow::Borrowed(line.text), line.eol));
                    continue;
                }
            } else if is_closing_fence(trimmed, fence_marker, fence_len) {
                in_fence = false;
                out.push((Cow::Borrowed(line.text), line.eol));
                continue;
            }
        }
        if in_fence {
            out.push((Cow::Borrowed(line.text), line.eol));
            continue;
        }

        // A pending paragraph immediately followed by a setext underline is a
        // heading; emit both verbatim so the underline keeps matching. (Checked
        // before prose classification because "===" is not otherwise a block.)
        if !para.is_empty() && is_setext_underline(line.text) {
            for l in &para {
                out.push((Cow::Borrowed(l.text), l.eol));
            }
            para.clear();
            out.push((Cow::Borrowed(line.text), line.eol));
            continue;
        }

        if is_prose(line.text) {
            para.push(*line);
            continue;
        }

        // Any non-prose line ends the current paragraph and passes through.
        flush(&mut out, &mut para, width);
        out.push((Cow::Borrowed(line.text), line.eol));
    }
    flush(&mut out, &mut para, width);

    let mut result = String::new();
    for (text, eol) in out {
        result.push_str(&text);
        result.push_str(eol);
    }
    result
}

/// Walks `root` recursively and wraps every `*.md` file in place to `width`
/// columns, rewriting only files whose content actually changes. Returns the
/// paths of the files it rewrote, in walk order.
pub fn wrap_directory(root: impl AsRef<Path>, width: usize) -> io::Result<Vec<PathBuf>> {
    let mut changed = Vec::new();
    visit_path(root.as_ref(), width, &mut changed)?;
    Ok(changed)
}

fn visit_path(path: &Path, width: usize, changed: &mut Vec<PathBuf>) -> io::Result<()> {
    let metadata = fs::metadata(path)?;
    if metadata.is_dir() {
        return visit_dir(path, width, changed);
    }
    if metadata.is_file() && is_markdown(path) {
        wrap_path(path, width, changed)?;
    }
    Ok(())
}

fn visit_dir(dir: &Path, width: usize, changed: &mut Vec<PathBuf>) -> io::Result<()> {
    let mut entries: Vec<_> = fs::read_dir(dir)?.collect::<io::Result<_>>()?;
    entries.sort_by_key(|entry| entry.path());

    for entry in entries {
        let path = entry.path();
        if entry.file_type()?.is_dir() {
            visit_dir(&path, width, changed)?;
        } else if is_markdown(&path) {
            wrap_path(&path, width, changed)?;
        }
    }
    Ok(())
}

fn wrap_path(path: &Path, width: usize, changed: &mut Vec<PathBuf>) -> io::Result<()> {
    let data = fs::read_to_string(path)?;
    let wrapped = wrap_str(&data, width);
    if wrapped != data {
        // Writing to the existing file keeps its permissions.
        fs::write(path, wrapped)?;
        changed.push(path.to_path_buf());
    }
    Ok(())
}

/// Wraps the pending paragraph (if any) into `out` and clears it.
fn flush<'a>(
    out: &mut Vec<(Cow<'a, str>, &'a str)>,
    para: &mut Vec<LogicalLine<'a>>,
    width: usize,
) {
    if para.is_empty() {
        return;
    }
    for (text, eol) in wrap_paragraph(para, width) {
        out.push((Cow::Owned(text), eol));
    }
    para.clear();
}

/// Greedily packs the words of a prose paragraph into lines no wider than `width`
/// display columns, never splitting a single word. Newly created lines reuse the
/// paragraph's first source eol; the final line keeps the last source eol so
/// trailing-newline and CRLF state survive.
fn wrap_paragraph<'a>(para: &[LogicalLine<'a>], width: usize) -> Vec<(String, &'a str)> {
    let mut words: Vec<&str> = Vec::new();
    for line in para {
        words.extend(line.text.split_whitespace());
    }
    if words.is_empty() {
        return para.iter().map(|l| (l.text.to_string(), l.eol)).collect();
    }

    let first_eol = para[0].eol;
    let last_eol = para[para.len() - 1].eol;

    let mut texts: Vec<String> = Vec::new();
    if width == 0 {
        texts.push(words.join(" "));
    } else {
        let mut cur = String::from(words[0]);
        let mut cur_w = disp_width(words[0]);
        for w in &words[1..] {
            let ww = disp_width(w);
            if cur_w + 1 + ww <= width {
                cur.push(' ');
                cur.push_str(w);
                cur_w += 1 + ww;
            } else {
                texts.push(std::mem::take(&mut cur));
                cur = String::from(*w);
                cur_w = ww;
            }
        }
        texts.push(cur);
    }

    let last = texts.len() - 1;
    texts
        .into_iter()
        .enumerate()
        .map(|(i, t)| (t, if i == last { last_eol } else { first_eol }))
        .collect()
}

/// Reports whether a line is ordinary paragraph text — it has content and is none
/// of the block constructs that must pass through untouched. Fenced code,
/// indented code, and front matter are handled by the caller before this.
fn is_prose(line: &str) -> bool {
    if line.trim().is_empty() || indented_too_far(line) {
        return false;
    }
    !(is_atx_heading(line)
        || is_blockquote(line)
        || is_list_item(line)
        || is_thematic_break(line)
        || is_html_block_start(line)
        || is_pipe_row(line))
}

/// Count of leading ASCII spaces, capped at the 4 that matter for CommonMark.
fn leading_spaces(line: &str) -> usize {
    line.bytes().take_while(|&b| b == b' ').take(4).count()
}

/// ATX heading: 1-6 `#` then a space or end of line, indented at most 3 columns.
fn is_atx_heading(line: &str) -> bool {
    if indented_too_far(line) {
        return false;
    }
    let b = &line.as_bytes()[leading_spaces(line)..];
    let hashes = b.iter().take_while(|&&c| c == b'#').count();
    if !(1..=6).contains(&hashes) {
        return false;
    }
    match b.get(hashes) {
        None => true,
        Some(&c) => c == b' ' || c == b'\t',
    }
}

/// Block quote: a `>` at most three columns in.
fn is_blockquote(line: &str) -> bool {
    if indented_too_far(line) {
        return false;
    }
    line.as_bytes().get(leading_spaces(line)) == Some(&b'>')
}

/// Bullet or ordered list item marker.
fn is_list_item(line: &str) -> bool {
    if indented_too_far(line) {
        return false;
    }
    let b = &line.as_bytes()[leading_spaces(line)..];
    if b.is_empty() {
        return false;
    }
    if matches!(b[0], b'-' | b'*' | b'+') {
        return b.len() == 1 || b[1] == b' ' || b[1] == b'\t';
    }
    let digits = b.iter().take(9).take_while(|c| c.is_ascii_digit()).count();
    if digits == 0 || digits >= b.len() {
        return false;
    }
    if b[digits] != b'.' && b[digits] != b')' {
        return false;
    }
    let rest = digits + 1;
    rest == b.len() || b[rest] == b' ' || b[rest] == b'\t'
}

/// Thematic break: three or more of a single `-`, `*`, or `_` with only spaces
/// between them.
fn is_thematic_break(line: &str) -> bool {
    if indented_too_far(line) {
        return false;
    }
    let s = line.trim();
    if s.len() < 3 {
        return false;
    }
    let marker = s.as_bytes()[0];
    if !matches!(marker, b'-' | b'*' | b'_') {
        return false;
    }
    let mut count = 0;
    for &b in s.as_bytes() {
        if b == marker {
            count += 1;
        } else if b != b' ' {
            return false;
        }
    }
    count >= 3
}

/// Setext heading underline: a run of only `=` or only `-`, indented ≤3 columns.
fn is_setext_underline(line: &str) -> bool {
    if indented_too_far(line) {
        return false;
    }
    let s = line[leading_spaces(line)..].trim_end_matches([' ', '\t']);
    if s.is_empty() {
        return false;
    }
    let marker = s.as_bytes()[0];
    if marker != b'=' && marker != b'-' {
        return false;
    }
    s.bytes().all(|b| b == marker)
}

/// HTML block start: a `<` at most three columns in.
fn is_html_block_start(line: &str) -> bool {
    if indented_too_far(line) {
        return false;
    }
    line.as_bytes().get(leading_spaces(line)) == Some(&b'<')
}
